### ISSUE TITLE
Introduce ability to run a pre-revocation code similar to how now we run `revoked`

### ISSUE THREAD

**Issue by mensfeld** [karafka/karafka#1413]
We need to hook up to the rebalance callbacks. This could be useful for mom scenarios late offset committing.

**Comment by lucastosetto** [comment]
Hey @mensfeld , I'd like to work on that

**Comment by mensfeld** [comment]
@lucastosetto this week I am traveling so if you don't mind I would get back to this next week.

Few things that need to happen:

1. There is a small inconsistency in the instrumentation as we already have `consumer.revoke` and `consumer.revoked` related to the same revocation flow in the default strategy. I can imagine that this could stay if the naming of pre-revoke would be different but just pointing out.
2. There are two stages of a rebalance: revoke and assign - I guess we would be interested in the pre-revoked phase in order to be able to commit offsets.
3. Naming - no idea atm how to name it

There is however alternative approach:

1. We do not use "lost_partitions" and always revoked even if partition is reassigned (we do it already).
2. Because of that we in theory could move the `#revoked` execution call to the revoked phase and introduce a new second callback `#assigned` (separate task) to allow to take actions on assignment prior to any data being available. This would allow us to have a full flow of assigned => consume => revoked.
3. What would be the side effect of "early" revoked on existing users? Not much. Anyhow any cleanup can be done as no processing is going to happen and the only thing that would change is the fact that `mark_as_consumed` would still work and would flush offsets afterwards.


I would opt on the second option because it would not require drastic instrumentation changes, would improve the current flow and would not change the API. But I also want to know your opinion.

**Comment by mensfeld** [comment]
Yeah, the more I look at it, the more I see that we could with fairly small effort move the revoked trigger to the revoked phase prior to final client offset commit. We would have to yield control back to run: 

```ruby
build_and_schedule_revoke_lost_partitions_jobs
wait
```

but that should not be a problem.

The only problem that I see now is a potential work saturation, if we have workers busy processing work from other consumer groups, this will hold WHOLE rebalancing. This means, that there could be a multi-consumer group race-condition. With the current approach there is no such risk because those "post revocation" jobs can wait as long as they want without affecting polling in any way (aside from blocking next poll). This means that the whole rebalance process at the moment is not blocked.

**Comment by mensfeld** [comment]
In theory, we could use the job `before_enqueue` API similar to how we control some of the advanced strategies flows. Since the hang of rebalance is inevitable due to the unpredictability of user code, I don't think we should worry about that. What we should worry however is the potential job saturation that would cause hang even **before** user code has a chance to kick in. `handle_before_enqueue` for LRJ does exactly that. It runs custom controller strategy hook for pausing from the main thread.

A HUGE downside would be to run unorganized third-party / user code from this thread because of low isolation. Hence if any I think we should run it FROM threads with saturation being documented. Alternatively we could introduce priority to the queue but they would have to be context aware.

**Comment by mensfeld** [comment]
What do I mean by context aware? Since we are in a rebalance callback, it means that given subscription group has no work to do (aside from LRJs). This means that this type of work could always be done first.

This problem (work saturation) may be the most visible for LRJ but it needs to be handled. Another solution would be to use a dedicated worker for things like this but I do not like that mostly because it starts differentiate jobs and may cause other problems.

**Comment by mensfeld** [comment]
One last idea: we could however skip the "pre revocation" job scheduling entirely unless it is explicitly configured, thus effectively running it and potentially saturating the flow only when **explicit** declaration of this flow occurs. This would basically not cause any saturation risks for LRJ, would run in worker threads (isolation) and would allow for anything. If this would be mixed with reorganization of the moment we run `#revoked`, we would then gain ability to run pre revocation operations in a robust way without significant risks.  Ofc if someone would put `sleep(1000000)` or threads would be heavily saturated, there would not be a way to get out of it and consumer would be kicked out of the CG.

I can imagine this could be easily configured per route on a topic level where an explicit declaration could be used (something like this):

```ruby
topic :my_topic do
  revoked_hook(:before) # :after as default
end
```

**Comment by mensfeld** [comment]
Btw @lucastosetto any chance of describing the use-case here? I'm just interested :)

**Comment by lucastosetto** [comment]
> this week I am traveling so if you don't mind I would get back to this next week.

No problem, I believe I wouldn't be able to finish the PR by the end of the week anyway.

> What we should worry however is the potential job saturation that would cause hang even before user code has a chance to kick in.

Thanks for pointing that out, now I'll keep an eye on it.

> One last idea: we could however skip the "pre revocation" job scheduling entirely unless it is explicitly configured

I'd better take a deeper at the code before confirming that, but this one sounds the most suitable approach at the moment. 

> any chance of describing the use-case here?

No specific problem I've faced, but I find it interesting to take advantage of this hook to persist state of unfinished tasks or to perform graceful disconnections from external dependencies.

**Comment by mensfeld** [comment]
@lucastosetto thanks for the aswer. I will lie the plan for this and sync up with you next week. :pray: 
there are few things that need to be done and happen to make sure that's robust around the edge cases.

**Comment by mensfeld** [comment]
@lucastosetto if you would join our community slack channel, this would be helpful: http://slack.karafka.io

**Comment by mensfeld** [comment]
@lucastosetto I wrapped my head around that I think.

We can sync next week :)

**Comment by mensfeld** [comment]
@lucastosetto I made the low-level changes needed for this to work here: https://github.com/karafka/karafka/pull/1662 

This should provide you with the layer for pre and post rebalance events. The only thing I am not sure (aside from syncing with you ofc ;) ) is the exact DSL for the routing.

**Comment by lucastosetto** [comment]
Hey @mensfeld , thanks for that! Just reached you out on Slack. Taking a look at the updates right now, I'll start a draft tomorrow and let you know.
